R/make_split_fun.R

```diff
@@ -0,0 +1,294 @@
+#' @param extra list. Extra arguments to be passed to analysis functions applied
+#' within the resulting facet.
+#'
+#' @details For `add_combo_facet`, the data associated with the resulting
+#' facet will be the data associated with the facets for each level in
+#' `levels`, rbound together. In particular, this means that if those levels
+#' are overlapping, data that appears in both will be duplicated.
+#'
+#' @return a function which can be used within the `post` argument in
+#' `make_split_fun`
+#'
+#' @seealso \code{\link{make_split_fun}}
+#'
+#' mysplfun <- make_split_fun(post = list(add_combo_facet("A_B", label = "Arms A+B",
+#'                                                        levels = c("A: Drug X", "B: Placebo")),
+#'                                        add_overall_facet("ALL", label = "All Arms")))
+#'
+#' lyt <- basic_table(show_colcounts = TRUE) %>%
+#'     split_cols_by("ARM", split_fun = mysplfun) %>%
+#'     analyze("AGE")
+#'
+#' tbl <- build_table(lyt, DM)
+#'
+#' @export
+#' @family make_custom_split
+add_combo_facet <- function(name, label = name, levels, extra = list()) {
+    function(ret, spl, .spl_context, fulldf) {
+        val <- LevelComboSplitValue(val = name, extr = extra, combolevels = levels, label = label)
+        add_to_split_result(ret, values = list(val), labels = label,
+                            datasplit = list(do.call(rbind, ret$datasplit[levels])))
+    }
+}
+
+#' @rdname add_combo_facet
+#' @export
+add_overall_facet <- function(name, label, extra = list()) {
+    add_combo_facet(name = name, label = label, levels = select_all_levels,
+                    extra = extra)
+}
+
+#' Trim Levels of Another Variable From Each Facet (Postprocessing split step)
+#' @param innervar character. The variable(s) to trim (remove
+#' unobserved levels) independently within each facet
+#'
+#' @return a function suitable for use in the `pre`
+#'     (list) argument of `make_split_fun`
+#' @seealso make_split_fun
+#' @export
+#' @family make_custom_split
+trim_levels_in_facets <- function(innervar) {
+    function(ret, ...) {
+        for(var in innervar) {
+            ret$datasplit <- lapply(ret$datasplit, function(df) {
+                df[[var]] <- factor(df[[var]])
+                df
+            })
+        }
+        ret
+    }
+}
+
+#' Preprocessing Functions for use in make_split_fun
+#'
+#' This function is intended for use as a preprocessing
+#' component in `make_split_fun`, and should not be called
+#' directly by end users.
+#'
+#' @param df data.frame. The incoming data corresponding with the parent facet
+#' @param spl Split.
+#' @param ... dots.
+#' @export
+#' @seealso make_split_fun
+#' @family make_custom_split
+drop_facet_levels <- function(df, spl, ...) {
+    if(!is(spl, "VarLevelSplit") || is.na(spl_payload(spl)))
+        stop("Unable to determine faceting variable in drop_facet_levels application.")
+    var <- spl_payload(spl)
+    df[[var]] <- factor(df[[var]])
```

Comment:
out of curiosity, why not
```suggestion
    df[[var]] <- droplevels(df[[var]])
```